File: src/main/java/net/magik6k/mpt/util/VersionUtil.java
package net.magik6k.mpt.util;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import net.magik6k.mpt.db.PackageBase;

import com.google.common.base.Charsets;

public class VersionUtil {
	private static MessageDigest md5;
	
	static{
		try {
			md5 = MessageDigest.getInstance("MD5");
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
	}
	
	public static String getNewHash(String pack){
		synchronized (md5) {
			md5.reset();
			md5.update(pack.getBytes(Charsets.UTF_8));
			return new BigInteger(md5.digest()).toString(16);
		}
	}
	
	public static String getPackageHash(String pack){
		synchronized (md5) {
			md5.reset();
			md5.update(pack.getBytes(Charsets.UTF_8));
			for(MptFile file : PackageBase.instance.getFiles(pack)){
				md5.update(file.name.getBytes(Charsets.UTF_8));
				md5.update(file.content.getBytes(Charsets.UTF_8));
			}
			return new BigInteger(md5.digest()).toString(16);
		}
	}
}

File: server/app/models/util/VersionUtil.scala
package models.util

import java.math.BigInteger
import java.security.MessageDigest

import com.google.common.base.Charsets
import models.Database

object VersionUtil {
  val md5 = MessageDigest.getInstance("md5")

  def getNewHash(pack: String): String = {
    md5 synchronized {
      md5.reset()
      md5.update(pack.getBytes(Charsets.UTF_8))
      new BigInteger(md5.digest()).toString(16)
    }
  }

  def getPackageHash(pack: String): String = {
    md5 synchronized {
      md5.reset()
      md5.update(pack.getBytes(Charsets.UTF_8))
      val files = Database.packageFiles(pack)
      files.foreach(f => md5.update(f.getBytes(Charsets.UTF_8)))
      files.map(f => Database.getPackageFile(pack, f)).foreach(f => md5.update(f.getBytes(Charsets.UTF_8)))
      Database.packageDependencies(pack).foreach(d => md5.update(d.getBytes(Charsets.UTF_8)))

      new BigInteger(md5.digest()).toString(16)
    }
  }
}
